### ISSUE TITLE
LSTM and internal state handling potentially not correct

### ISSUE THREAD

**Issue by AlexKuhnle** [tensorforce/tensorforce#128]
Currently, models are trained per-timestep, i.e. one batch instance is one timestep. Hence explicitly they learn to optimize the output per-timestep, but not the conditional dependence of timesteps within-episode introduced by internal states.

A solution could be to instead consider short successive state/action sequences as batch items, but this would explicitly only cover short-distance dependencies well. Is this problematic?

**Comment by trickmeyer** [comment]
Thanks @AlexKuhnle.  From my understanding, it's industry standard to truncate the theoretically infinite length sequences down to something manageable.  For example, [Tensorflow's Recurrent section](https://www.tensorflow.org/tutorials/recurrent) puts it nicely:

Truncated Backpropagation

By design, the output of a recurrent neural network (RNN) depends on arbitrarily distant inputs. Unfortunately, this makes backpropagation computation difficult. In order to make the learning process tractable, it is common practice to create an "unrolled" version of the network, which contains a fixed number (num_steps) of LSTM inputs and outputs. The model is then trained on this finite approximation of the RNN. This can be implemented by feeding inputs of length num_steps at a time and performing a backward pass after each such input block.


Different papers handle the fact that you're "starting from zero" at the start of each sequence differently.  For example, one [successful approach to playing Doom developed at CMU](https://arxiv.org/pdf/1609.05521.pdf) sent only the last half of the gradients for a given trace backward instead of sending all the gradients backwards when training their agent.  They did so by recognizing that the initialization that occurs in the first few steps of a new sequence is likely to be noisy.  This, along with a few other tricks, allowed them to achieve state of the art performance.  Here's a "one pager" [implementation](https://github.com/awjuliani/DeepRL-Agents/blob/master/Deep-Recurrent-Q-Network.ipynb) trying to recreate those results as well as an accompanying [blog post](https://medium.com/emergent-future/simple-reinforcement-learning-with-tensorflow-part-6-partial-observability-and-deep-recurrent-q-68463e9aeefc).

**Comment by lefnire** [comment]
@AlexKuhnle check out https://github.com/awjuliani/DeepRL-Agents/commit/16ebef4bc36689c2d1859543b8c23aec8cda3a99 to fix just that in his repo - stores RNN state along w/ current state replay memory. Treating it as "just part of the state" gives it that Markovian goodness that Dense layers have.

**Comment by AlexKuhnle** [comment]
@lefnire  Yes, that's what currently happens (right?). Do they also sample short sequence snippets from the memory?

**Comment by trickmeyer** [comment]
@AlexKuhnle @lefnire that's the same link I sent across in the Gitter last week where consecutive sequences of 8 are sampled from the memory (they're stored by episode in this implementation), and gradients are only sent back on the latter half of cases to account for the "initialization noise" of these shorter sequences.

**Comment by trickmeyer** [comment]
This is different than how things currently work re: our Gitter conversations.  @AlexKuhnle you and I came up with some ways to go about changing this.  I hope to get to them in early October and use an implementation similar to the link @lefnire sent as a benchmark.  @lefnire - if you have time to create a branch before then, check out the conversation in Gitter from earlier this week and please pass along the work for us to check out!

**Comment by AlexKuhnle** [comment]
Sorry, should have been more precise -- when I said it's what currently happens, I was only referring to the part "...stores RNN state along w/ current state in replay memory".

**Comment by AlexKuhnle** [comment]
Actually, just a small remark if you guys start working on it. As I mentioned earlier, I'm a bit reluctant to change the batch format from [batch, timesteps] to a 3-rank shape [batch, episodes, timesteps]. Again, we mentioned before that one way could be to just locally change it in the LSTM layer. However, I wonder whether a better solution would be to write our own recurrent module operating as we want, that is, reading N-timestep-subsequences from the batch sequence of timesteps, stopping if a terminal is encountered, potentially applying tf.stop_gradient after some steps, etc. That might give us lots of flexibility and potential for configuration.

**Comment by trickmeyer** [comment]
@AlexKuhnle I'm still getting to the point where the Tensorforce interfaces/architecture is second-nature to me, so it's difficult for me to opine on the approach at this point.  Before I do any work on this I'll do a deep dive and try to compare the two different approaches.  In the meantime, if you're interested, it'd be nice to have an "inventory" / list of the work to be done with each approach to do a fair comparison of the two approaches.

**Comment by AlexKuhnle** [comment]
No worries, this was more just a thought that I wanted to record somewhere...  ;-)  Definitely not directly addressing the other implementation, which I haven't yet looked at in depth, but I should do at some point to be able to compare better (if they even are different).

**Comment by ervteng** [comment]
Apologize for bringing up an old thread - will the new Memory branch resolve this issue? As my project requires the LSTM playback to work properly, I'm debating on whether I should implement sequence playback for the old branch (which I can get my head around =P) or switching to the new branch. Thanks!!

**Comment by AlexKuhnle** [comment]
Hi @ervteng, no problem, since this is still an existing issue. :-/  The memory branch implements the (so far little tested) ability to retrieve short temporally consistent sequences from memory (like [here](https://github.com/reinforceio/tensorforce/blob/memory/tensorforce/core/memories/replay.py#L57)), or entire episodes in consistent order. In that sense there is progress in the new branch, however, it is not (yet) properly integrated with appropriate learning of internal state modules. But it is probably worth starting from there, also more likely to result in a PR, if you start working on this.

**Comment by trickmeyer** [comment]
@AlexKuhnle i've written a "trace sampler" that grabs temporally-consistent sequences that don't "overlap" like the sequence sampler, but could use some help on what it would take to, in your words, get it "properly integrated with appropriate learning of internal state modules."  Could you add a bit more details on what's required?  I'd love to get LSTM with PPO working, there's some interesting work being done in this area.

**Comment by AlexKuhnle** [comment]
Hi @trickmeyer, totally agree that this would be very good to have. Here a few thoughts: We essentially have three modes now, timestep-based (not useful for anything sequential), episode-based and subsequence-based (which is, in a way, timestep-based but compatible with sequential modules like LSTM). The subsequence-based mode basically has a state batch format of `(None, seq_length, *state_shape)` and the LSTM iterates over the first dimension (note that sequences seem only necessary for states) before outputting an action (instead of the usual direct state-to-action). For the episode-based mode the batch format is `(None, *state_shape)`, but different from the timestep-based mode here the LSTM can actually iterate over the entire sequence of states, dimension 0 (however, resetting at every terminal). Note that an actual batch size of `(None, seq_length, ...)` would not work with per-timestep dense layers etc, so instead a `sequence_length` parameter is required, saying how the `(None, ...)` batch format is to be interpreted. Three options: 1 (timestep-based), n (subsequence-based, terminal information might be helpful), * or so (episode-based, requires terminal information). Does that give an idea of how the different modes could come together and what is required for such a solution? Might well be I miss something...

**Comment by trickmeyer** [comment]
@AlexKuhnle thanks for the writeup.  From my understanding of LSTM requirements, we need two properties for proper sampling:

1.  Finite-length, temporally-ordered sequences of state samples, and
2. the sequences shouldn't "span" episodes


From my reading of the current code, it seems like we have three `Memory` sampling methods:

1. `tf_retrieve_timesteps`
2. `tf_retrieve_episodes`
3. `tf_retrieve_sequences` (i believe you referred to as subsequence)


Assuming all the above is correct, we can't use `tf_retrieve_timesteps` in any of the `core/memory` classes because it doesn't limit sampling to within episodes (e.g. even `latest.py` can span episodes).

I actually don't believe `tf_retrieve_sequences` is an intuitive implementation for LSTM models, although perhaps it could be wrangled to be correct with a postprocess.  

I think the core issue is it essentially being a "concatenating convolution" over the states most users are putting in as the independent observations, which changes the state shape.  For example, if I have an extremely simple state with values [1], [2], ..., [n] at timesteps 1, 2, ..., n, my reading of the `tf_retrieve_sequences` code with sequence length 8 would return `[[1], [2], ..., [8]]`, `[[2], [3], ..., [9]]`, `[[3], [4], ..., [10]]` as three observations, basically changing the input size from `(None, 1, )` to `(None, 8, 1)`.  

In reality, I think most users would want ordered samples of size `(1, )` 8 at a time, and not spanning multiple episodes.  For that reason `tf_retrieve_episodes` may very well work, but the issue is that it returns entire episodes, and many LSTM training schemes I've seen use random "intra-episode" samples of size 8 (to keep the example going) as an approximation of the infinite sequence.  Given the current implementation, that means I could limit my episodes to length 8 with `tf_retrieve_episodes` to get observations flowing appropriately, but perhaps my environment needs longer episodes to experience what it needs to in order to figure out a good policy.

Ideally, I'd add a `tf_retrieve_traces` method ("trace" is a word I've seen with LSTMs) with a pretty simple algorithm that is just a minor modification of `tf_retrieve_episodes`:

```
def tf_retrieve_traces(trace_length, num_traces):
  * assume we have constant episode length of episode_length
  * get all episode start, stop indices similar to `tf_retrieve_episodes`
  * for episode in episodes:
    * pick random start index i in [0, episode_length - trace_length - 1]
    * yield trace slice from [episode.start_index + i, episode.start_index +i + trace_length]
```

This way, the input state shape is preserved, we get "long episodes" to discover proper dynamics, we get a temporally-consistent sequence of those preserved state shapes, and we make sure all samples are taken from a single episode.

My followup question relates to what happens "downstream" from the `Memory` sampling, but based on my first pass of the new codebase, it seems it passes internals correctly if we pass in states correctly with the proposed scheme.  It would be nice to get someone with more significant LSTM experience to chime in and confirm that downstream handling is proper at some point, but I think this would be a great start.  I have the code written up, but it does require a loop over each episode to get eligible random start/stop indices, since tf.random_uniform can't be used unless we have a constant episode length.

**Comment by AlexKuhnle** [comment]
Note that the `tf_retrieve_sequences` method is not really in use yet, but meant to behave like your `tf_retrieve_traces`. Whether the batch format should be `(None, *state_shape)` or `(None, trace_len, *state_shape)` depends on the downstream mechanism, but I agree that the former is likely a better solution, and `tf_retrieve_sequences` hence might need changing. Also I'm fine with renaming it to what you propose. I don't think fixed episode length is necessary (`Queue` stores terminal positions), and I don't think traces have to necessarily come from the same episode (since batch instances are usually assumed independent, why should they?).

Downstream processing has to somehow be handled by the LSTM (or other internal-state) module(s). I mentioned a `sequence_length` parameter as an option to handle it. After the internal state module, only the last output will be passed on, I assume, so from then on within the network the batch looks as before, and subsequent modules are not affected. So I think it's only the internal-state layers and general handling in the network, that needs to be modified. What do you think?

**Comment by trickmeyer** [comment]
Regarding observing/sampling, I believe having sequential "traces" from the same episode is required.  See this [Tensorflow overview](https://www.tensorflow.org/tutorials/recurrent) of RNNs for a good simple example and code.  You'll also see that they do have traces of a constant length, but I do see that it's common to pad if some samples are shorter.

Regarding downstream work, I think it'd be great if we maybe reached out on the dev chat or something to try to see if we have any RNN experts that are participating to see if they could weigh in.

What do you think?

**Comment by AlexKuhnle** [comment]
Each trace has to be from one episode, of course, but a batch can contain traces each from different episodes, I'm pretty sure. A sequence length argument can be specified for an LSTM, but for us the only case where padding is required is at the start, and there it probably makes sense to just pad with the start state (or use the sequence length argument, which might be the cleanest solution).

If someone with RNN-in-RL experience is around, that would be great, of course, to confirm what we are planning (and/or actively engage). What do you mean by "downstream work", tasks to evaluate? Yes, that would also be helpful.

**Comment by trickmeyer** [comment]
Yes, completely agreed: individual trace from individual episode, individual batch from (likely) many traces/episodes.  Sorry I misunderstood your earlier comment.

Concerning sampling traces of a constant length, it seems we have three options:

1. Only sample from episodes with length >= minimum.
2. Pad beginning.
3. Pass trace length along with trace.

I think it makes sense to add the `tf_retrieve_(something)` method to memory, and perhaps include the option as a parameter.  We could start with (1) from above pretty easily, and I also believe it'd be the most common use case.  The others could be implemented others later.

Once we get that I'm confident in our sampling.  But concerning downstream, I really want to check the areas I'm not as familiar with in both the current tensorforce code as well as in RNN/RL theory.  Mainly two areas:

1. Are we passing internals correctly?
2. Are we resetting at the right times, and are all other `observe` "lifecycle" operations currently implemented fine as-is?


I can submit a PR for the `tf_retrieve_(something)` - what should we call it?  I don't know if anyone is using `tf_retrieve_sequences` right now, and I do think that method has some merit as it's currently implemented for other use cases (frame stacking, etc).

**Comment by AlexKuhnle** [comment]
Sorry, I missed this last post, but sounds all fine. The `retrieve_sequences` shouldn't really be in use so far (it's not really compatible), and was specifically added to support this feature in the future, so I think you can use and modify it (e.g. add another argument). So far I would think that frame stacking is supposed to happen as `sequence` preprocessing.

**Comment by AlexKuhnle** [comment]
Outdated
